### ISSUE TITLE
after upgrade HA to 99.3

### ISSUE THREAD

**Issue by ashrafmn** [peterbuga/HASS-sonoff-ewelink#145]
my sonoff stoped working after upgrade 99.3
getting the below errors

2019-09-27 14:24:48 ERROR (MainThread) [homeassistant.setup] Error during setup of component sonoff
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/setup.py", line 168, in _async_setup_component
    hass, processed_config
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 53, in async_setup
    hass.data[DOMAIN] = Sonoff(hass, config)
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 108, in __init__
    self.do_login()
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 192, in do_login
    self.update_devices() # to write the devices list
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 357, in update_devices
    if self.is_grace_period():
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 322, in is_grace_period
    grace_time_elapsed = self._skipped_login * int(SCAN_INTERVAL.total_seconds())
NameError: name 'SCAN_INTERVAL' is not defined

**Comment by peterbuga** [comment]
downgrade then

**Comment by ashrafmn** [comment]
@peterbuga @difelice 
super,, i did the modification in __init__.py file and my sonoffs are back now...
thnx a lot for the prompt action,, and i didn't need to downgrade!!

**Comment by ashrafmn** [comment]
i noticed that the switches state keep changing to "Unavailable" rapidly 

and the below are reported in the log

2019-09-28 12:09:59 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 266, in async_update_ha_state
    self._async_write_ha_state()
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 300, in _async_write_ha_state
    state = self.state
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 554, in state
    return STATE_ON if self.is_on else STATE_OFF
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/switch.py", line 75, in is_on
    self._state = self.get_state()
  File "/home/homeassistant/.homeassistant/custom_components/sonoff/__init__.py", line 649, in get_state
    return device['params']['switch'] == 'on' if device else False
KeyError: 'switch'
2019-09-28 12:10:09 WARNING (MainThread) [custom_components.sonoff] Grace period activated!
2019-09-28 12:11:20 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:11:20 ERROR (Thread-8) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:11:24 ERROR (Thread-8) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:11:24 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:11:34 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:11:34 ERROR (Thread-8) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x68512e10>>: string indices must be integers
2019-09-28 12:12:00 WARNING (MainThread) [custom_components.sonoff] Grace period activated!

**Comment by peterbuga** [comment]
@ashrafmn post `sonoff-debug.py` output, i believe you might have a sonoff B1 (or something similar) in your devices list

**Comment by ashrafmn** [comment]
@peterbuga file uploaded


[devices_28Sep.json.txt](https://github.com/peterbuga/HASS-sonoff-ewelink/files/3665303/devices_28Sep.json.txt)

**Comment by peterbuga** [comment]
@ashrafmn thanks for the dump 👍, looks like I was right. 
unfortunately the above fix was made on `master` branch and i'm pretty sure you were on `sonoff-bulb` branch when you reported this problem. 

TL;DR: `sonoff-bulb` not working for you right now unless you manually edit it with the patches from current `master` **OR** use `master` branch but remove the sonoff B1s devices **OR** wait till i'll merge `sonoff-bulb` into `master` 😬

**Comment by ashrafmn** [comment]
@peterbuga man you are doing extremely good job here,, 
i downloaded the master branch,, and now I'm using it as is,, looks like the devices states are more stable now and no more "Unavailable" state like before

looking forward to fix this in all branches in addition to my other request "sonoff sc #74"

thnx again

**Comment by dannygohks** [comment]
> @peterbuga @difelice
> super,, i did the modification in **init**.py file and my sonoffs are back now...
> thnx a lot for the prompt action,, and i didn't need to downgrade!!

what do you do? i also need to fixed it.

**Comment by peterbuga** [comment]
@dannygohks download `master` branch if you don't have Sonoff B1s in your devices list, this problem has been fixed.

**Comment by difelice** [comment]
Hi @peterbuga, I think there's still a problem around `scan_interval` and `grace_period`, at least on my case `homeassistant==0.99.3` and `master`:

## History Chart
![image](https://user-images.githubusercontent.com/3169762/65832704-0eea1a00-e28d-11e9-9e69-7b2512134dd7.png)

## Log
```log
2019-09-29 13:40:17 INFO (MainThread) [custom_components.sonoff] Found websocket address: as-pconnect1.coolkit.cc
2019-09-29 13:40:17 DEBUG (Thread-3) [custom_components.sonoff] (re)init websocket
2019-09-29 13:40:19 DEBUG (Thread-3) [custom_components.sonoff] websocket msg: {"error":0,"apikey":"cdca5c4a-df3e-49bc-af84-7376fccd4541","config":{"hb":1,"hbInterval":145},"sequence":"1569760818823824"}
2019-09-29 13:40:29 WARNING (MainThread) [custom_components.sonoff] Grace period activated!
2019-09-29 13:41:18 INFO (MainThread) [custom_components.sonoff] Grace period active
2019-09-29 13:42:19 INFO (MainThread) [custom_components.sonoff] Grace period active
2019-09-29 13:43:20 INFO (MainThread) [custom_components.sonoff] Grace period active
2019-09-29 13:44:21 INFO (MainThread) [custom_components.sonoff] Grace period active
```

I'll try to setup a development environment since I had to downgrade both HA and sonoff component for my house.

Thanks.

**Comment by peterbuga** [comment]
this issue looks quite inconsistent, no big change was done than to move a variable around, all the previous logic is still there. 
are you using a 2nd ewelink account? if not, stop accessing the  ewelink app :P
in the same time, as a test disable if you have `debug: true` in the sonoff configuration and set `scan_interval: 60` (or remove it) this should put everything in sync (pooling time & grace period calculation)

ps: i'm sorry but the history chart doesn't tell me much, are those the sonoff devices?

**Comment by dannygohks** [comment]
I have updated the master 3 times, no more error but the devices is not sync into my HA. No error message except the info below.

Grace period activated!
10:43 AM custom_components/sonoff/__init__.py (WARNING) - message first occured at 10:37 AM and shows up 2 times

I'm using a newly created 2nd account. 

`sonoff:
  username: !secret ewelink_username
  password: !secret ewelink_password
  scan_interval: 120
  grace_period: 600
  api_region: 'as'
  entity_prefix: True
  debug: False`

**Comment by dannygohks** [comment]
I have updated the master 3 times, no more error but the devices is not sync into my HA. No error message except the info below.

Grace period activated!
10:43 AM custom_components/sonoff/init.py (WARNING) - message first occured at 10:37 AM and shows up 2 times

I'm using a newly created 2nd account.

![Screenshot_211](https://user-images.githubusercontent.com/48622859/65846274-ce170180-e36f-11e9-8b96-66af87c1ff54.png)

**Comment by difelice** [comment]
> this issue looks quite inconsistent, no big change was done than to move a variable around, all the previous logic is still there.

Yes, in my update I'm trying to explain that it could be related to ewelink API.
> are you using a 2nd ewelink account? if not, stop accessing the ewelink app :P

I'm not using the ewelink app.
> in the same time, as a test disable if you have `debug: true` in the sonoff configuration and set `scan_interval: 60` (or remove it) this should put everything in sync (pooling time & grace period calculation)

I have `scan_interval: 60`. Enabling `debug` does help, isn't that just a shorter `scan_interval`?
> 
> ps: i'm sorry but the history chart doesn't tell me much, are those the sonoff devices?

Yes, those are all Sonoff devices that started to look like that 3 days ago. I did update `homeassistant==0.90.3` but even after downgrading to `0.98` the issue persisted.

I'll try to get more detailed logs of the API responses.

Thanks for replying!

**Comment by peterbuga** [comment]
> Enabling debug does help, isn't that just a shorter scan_interval?

yes and no, after a while depending on the total number of components you have activated can overload HA and cannot successfully complete the pooling requests. all the updates are done via websocket which way more faster, `scan_interval` it's more like a fallback to be sure the devices get update anyway in case websocket fails (which happens) in particular moments.

if anything, go the opposite way and increase `scan_interval: 600` and let me know how it behaves. 

just like in the case of POW R2 fw 3.3.0+ problem, i think itead starts to be aware of this component and they are trying to limit the use of it because it uses *a lot* of bandwidth. they cannot change the anything related to login part because it's hardcoded in all the apps (🤦‍♂), lucky for us tho', instead they can add rules so my take on it is that if they see an account logged/requesting for a long period of time they might invalidate the session, `grace period` is triggered and repeat

**Comment by difelice** [comment]
@peterbuga thanks for the explanation! 

I'll keep debugging the problem on my end. I also noticed that we could activate the grace period only if devices were successfully pulled on first attempt.

Looks like HA does not register them and in my case I have to restart server every time to fix that.

A possible fix for that: https://github.com/peterbuga/HASS-sonoff-ewelink/pull/151

**Comment by peterbuga** [comment]
on average ~67 requests/sec sounds a bit ... not believable (as a side effect you'll see some high cpu usage too)
